aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -38,12 +38,12 @@ static inline Tensor integer_upcast(const Tensor& self, optional<ScalarType> dty
 
 using DimMask = TensorIterator::DimMask;
 
-static DimMask make_dim_mask(IntArrayRef dims, int64_t ndim) {
+static DimMask make_dim_mask(c10::optional<IntArrayRef> dims, int64_t ndim) {
   auto mask = DimMask();
-  if (dims.empty()) {
+  if (!dims.has_value() || dims.value().empty()) {
```

Comment:
dim will not have value if user do not specify any dim at all
dim will be empty if dim=[]

Reply:
I pointed @umanwizard at this -- my argument was roughly:
1) if these things do different things, they should do different things
2) if they do the same thing, we should have reduced the input space before this to only support one of them.

I think based on numpy behavior these should actually do different things, i.e.:

```
>>> np.var(np.random.randn(2,3), axis=())
array([[0., 0., 0.],
           [0., 0., 0.]])

>>> np.var(np.random.randn(2,3), axis=None)
1.6819117021242187
```

but I'll let @umanwizard make a call on that.

Reply:
After speaking with Greg and Sam, we agreed that there is no good reason for `dim=[]` to be interpreted as `dim=None`.

`sum` and `prod` are a better example of what's happening now. They *do* use TensorIterator for the no-dims case, unlike `std` and `var`. And the versions without dim just pass through an empty list of dims:

```
Tensor sum(const Tensor &self, ScalarType dtype) {
  return at::native::sum(self, {}, false, optional<ScalarType>(dtype));
}
```

Notice it's passing `{}` as an empty list of dims. So I guess that's why we are currently interpreting an empty list of dims as "all dims".

I think the best way to fix this is to have `make_reduction` take an `optional<IntArrayRef>`, and move the logic to check if it's null into there. Then we could remove this if check from `make_dim_mask` entirely, and have `make_dim_mask` return an empty mask if given an empty set of dims. We would then have to change sum, prod, norm, and whatever other functions we have that use TensorIterator, to pass a null optional when dims isn't provided, rather than pass an empty list.

I don't think it is necessary to do in this PR, though. It's fine for now if you just do the same thing as sum/prod do now (check out ReduceOps.cpp), and not change `make_dim_mask`. Later on after you land this I can make the necessary change to have empty dims work the same way as numpy.